_posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd

```diff
@@ -0,0 +1,705 @@
+
+```{r}
+train_ds <- brainseg_dataset(
+  train_dir,
+  augmentation_params = c(0.05, 15, 0.5),
+  random_sampling = TRUE
+)
+
+length(train_ds)
+# 2977
+
+valid_ds <- brainseg_dataset(
+  valid_dir,
+  augmentation_params = NULL,
+  random_sampling = FALSE
+)
+
+length(valid_ds)
+# 952
+```
+
+As a correctness check, let's plot an image and associated mask:
+
+```{r}
+par(mfrow = c(1, 2), mar = c(0, 1, 0, 1))
+
+img_and_mask <- valid_ds[27]
+img <- img_and_mask[[1]]
+mask <- img_and_mask[[2]]
+
+img$permute(c(2, 3, 1)) %>% as.array() %>% as.raster() %>% plot()
+mask$squeeze() %>% as.array() %>% as.raster() %>% plot()
+```
+
+```{r, eval = TRUE, echo = FALSE, layout="l-body-outset"}
+knitr::include_graphics("images/scan_and_mask.png")
+```
+
+With `torch`, it is straightforward to inspect what happens when you change augmentation-related parameters. We just pick a pair from the validation set, which has not had any augmentation applied as yet, and call `valid_ds$<augmentation_func()>` directly. Just for fun, let's use more "extreme" parameters here than we do in actual training. (Actual training uses the settings from Mateusz' GitHub repository, which we assume have been carefully chosen for optimal performance.[^1])
+
+[^1]: Yes, we did a few experiments, confirming that more augmentation isn't better ... what did I say about inevitably ending up doing optimization on the validation set ...
+
+```{r}
+img_and_mask <- valid_ds[77]
+img <- img_and_mask[[1]]
+mask <- img_and_mask[[2]]
+
+imgs <- map (1:24, function(i) {
+  
+  # scale factor; train_ds really uses 0.05
+  c(img, mask) %<-% valid_ds$resize(img, mask, 0.2) 
+  c(img, mask) %<-% valid_ds$flip(img, mask, 0.5)
+  # rotation angle; train_ds really uses 15
+  c(img, mask) %<-% valid_ds$rotate(img, mask, 90) 
+  img %>%
+    transform_rgb_to_grayscale() %>%
+    as.array() %>%
+    as_tibble() %>%
+    rowid_to_column(var = "Y") %>%
+    gather(key = "X", value = "value", -Y) %>%
+    mutate(X = as.numeric(gsub("V", "", X))) %>%
+    ggplot(aes(X, Y, fill = value)) +
+    geom_raster() +
+    theme_void() +
+    theme(legend.position = "none") +
+    theme(aspect.ratio = 1)
+  
+})
+
+plot_grid(plotlist = imgs, nrow = 4)
+```
+
+```{r, eval = TRUE, echo = FALSE, layout="l-body-outset"}
+knitr::include_graphics("images/brainseg_augmentation.png")
+```
+
+Now we still need the data loaders, and then, nothing keeps us from proceeding to the next big task: building the model.
+
+```{r}
+batch_size <- 4
+train_dl <- dataloader(train_ds, batch_size)
+valid_dl <- dataloader(valid_ds, batch_size)
+
+```
+
+## Model
+
+Our model nicely illustrates the kind of modular code that comes "naturally" with `torch`. We approach things top-down, starting with the U-Net container itself.
+
+`unet` takes care of the global composition -- how far "down" do we go, shrinking the image while incrementing the number of filters, and then how do we go "up" again?
```

Comment:
Are you referring to the function you create below? If so, `unet()` gets trailing parentheses.

Reply:
Oh, maybe it's an object you're creating, and not a function (in which case, ignore my comment above).

Reply:
yeah, i know ... i was thinking about that ... technically it's a constructor, so more like a function, but the semantics should really be noun-like ...